### ISSUE TITLE
bug: cluster-mode val split is sliced by num_shards in precompute_teacher.py, not run in full on shard-0

### ISSUE THREAD

**Issue by travisgalloway** [travisgalloway/monica#174]
Part of #65

## Problem

`scripts/precompute_teacher.py`'s cluster-mode chunk math applies uniformly to **every split**
passed via `--splits`, including `val`. The RunPod shard launcher (`scripts/runpod/m10/run_precompute_shard.sh`)
only assigns `val` to shard 0 (`SPLITS_ARG="train,val"` when `SHARD_ID -eq 0`, else `"train"`), on
the assumption that shard 0 then precomputes the **entire** val split. It doesn't:

```python
if cluster:
    per_shard, remainder = divmod(total_chunks, num_shards)
    start_chunk = shard_id * per_shard + min(shard_id, remainder)
    this_chunks = per_shard + (1 if shard_id < remainder else 0)
else:
    start_chunk, this_chunks = 0, total_chunks
```

`cluster = shard_id is not None` is set once per invocation and applies inside the `for split in
splits:` loop regardless of which split is being processed. So when shard 0 processes `val` with
e.g. `--num-shards 4`, `total_chunks` here is `val`'s own (small) chunk count, and shard 0 only
gets `per_shard = total_chunks // 4` of it — a quarter of val, not the whole thing.

## Impact — already manifested

This is not theoretical: the completed Phase-B precompute run (566 GB teacher-outputs merged to
`s3://monica-training/poc-distill/teacher-outputs/topk-logits-merged/`) has **val = 305 chunks**,
not the expected **1220** (per `docs`/memory notes on the `split --val-tokens 10000000` output).
305 ≈ 1220 // 4, consistent with a 4-shard cluster run hitting exactly this bug.

It hasn't caused silent corruption because `DistillLoader` asserts
`meta["n_chunks"] == packed_val_n_chunks` (`src/data/teacher_outputs.py:152`) — so this will
**hard-fail Phase C** the moment `distill.py` constructs a val `DistillLoader` (needed for
`--eval-every`), not silently train on incomplete data. But it does block the M10 milestone's
next phase (the two-layout sweep) until fixed.

## Fix

`val` should never be shard-divided — shard 0 should process it as if `cluster=False`. Two
reasonable approaches:
1. In `precompute_teacher.py`, special-case `split == "val"` to always use
   `start_chunk, this_chunks = 0, total_chunks` regardless of `cluster`/`shard_id`, OR
2. Have the shard launcher invoke `precompute_teacher.py` for `val` as a **separate**, non-cluster
   call (no `--shard-id`/`--num-shards`) on shard 0 only, then merge as today.

(1) is likely simpler and keeps the launcher's single invocation. Whichever approach, re-run val
precompute to get the full 1220 chunks and re-merge (`scripts/merge_teacher_shards.py` already
handles val as a shard-0-only passthrough as of #162 — that logic is unaffected by this fix, it
just needs a correctly-sized shard-0 val artifact to pass through).

## Task checklist
- [x] Fix `precompute_teacher.py`: `val` always uses `start_chunk, this_chunks = 0, total_chunks`
      regardless of `cluster`/`shard_id`.
- [x] Add a functional test: `--splits val --shard-id 0 --num-shards N` (N>1) → val meta
      `n_chunks` equals the full val chunk count, not `total_chunks // N`.
- [ ] (Pod-gated, NOT part of this PR) re-run real val precompute, re-merge, confirm
      `teacher-val.meta.json n_chunks == 1220`.

## Priority

**P0** — blocks Phase C (the two-layout distillation sweep, next after Phase B′) at the very first
`DistillLoader` construction for val. Needs a fix + a val-only re-precompute (cheap — val is only
~1220 chunks / 10M tokens, not a multi-day job) before Phase C launches.

## Verification
- Unit/functional: rerun `precompute_teacher.py --splits val --shard-id 0 --num-shards N` (N>1) on
  a synthetic/toy corpus and assert `n_chunks` in the output meta equals the full val chunk count,
  not `total_chunks // N`.
- Re-run the real val precompute on the existing corpus, re-merge, and confirm
  `teacher-val.meta.json n_chunks == 1220` before Phase C's C.3 preflight check.
